src/sentry/integrations/jira/utils/create_issue_schema_transformers.py

```diff
@@ -0,0 +1,141 @@
+    return input_val
+
+
+def id_obj_transformer(input_val: Any) -> dict[str, Any]:
+    return {"id": input_val}
+
+
+def get_type_transformer_mappings(user_id_field: str) -> TransformerType:
+    transformers = {
+        JiraSchemaTypes.user.value: lambda x: {user_id_field: x},
+        JiraSchemaTypes.issue_type.value: id_obj_transformer,
+        JiraSchemaTypes.option.value: lambda x: {"value": x},
+        JiraSchemaTypes.issue_link.value: lambda x: {"key": x},
+        JiraSchemaTypes.project.value: id_obj_transformer,
+        JiraSchemaTypes.number.value: parse_number_field,
+    }
+
+    return transformers
+
+
+def get_custom_field_transformer_mappings() -> TransformerType:
+    transformers = {
+        # JIRA_CUSTOM_FIELD_TYPES["select"]: identity_transformer,
+        JIRA_CUSTOM_FIELD_TYPES["tempo_account"]: parse_number_field,
+        JIRA_CUSTOM_FIELD_TYPES["sprint"]: parse_number_field,
+        # JIRA_CUSTOM_FIELD_TYPES["epic"]: identity_transformer,
+        JIRA_CUSTOM_FIELD_TYPES["rank"]: id_obj_transformer,
+    }
+
+    return transformers
+
+
+def get_transformer_for_field(
+    type_transformers: TransformerType, custom_transformers: TransformerType, jira_field: JiraField
+) -> Callable[[Any], Any]:
+    transformer = None
+    if jira_field.is_custom_field():
+        assert jira_field.schema.custom
+        transformer = custom_transformers.get(jira_field.schema.custom)
+
+    if not transformer:
+        field_type = jira_field.get_field_type()
+
+        if field_type:
+            transformer = type_transformers.get(field_type)
+
+    if not transformer:
+        transformer = identity_transformer
+
+    return transformer
+
+
+def transform_fields(
+    user_id_field: str, jira_fields: Iterable[JiraField], **data
+) -> Mapping[str, Any]:
+    transformed_data = {}
+
+    # Special handling for fields that don't map cleanly to the transformer logic
+    data["summary"] = data.get("title")
+    if labels := data.get("labels"):
+        data["labels"] = [label.strip() for label in labels.split(",") if label.strip()]
+
+    type_transformers = get_type_transformer_mappings(user_id_field)
+    custom_field_transformers = get_custom_field_transformer_mappings()
+
+    for field in jira_fields:
+        field_data = data.get(field.key)
+
+        # We don't have a mapping for this field, so it's probably extraneous
+        if field_data is None:
+            continue
+
+        field_transformer = get_transformer_for_field(
+            type_transformers, custom_field_transformers, field
+        )
+
+        try:
+            # We need to handle array types, but for whatever reason `Sprint`
+            # fields have a type of array/json, but expect a single value, so we
+            # have to special case them.
+            if (
+                field.schema.schema_type.lower() == JiraSchemaTypes.array
+                and field.schema.custom != JIRA_CUSTOM_FIELD_TYPES["sprint"]
```

Comment:
This is a really annoying edge case. The input type is different than the schema type. It expects a single sprint ID as an int, but returns a list of sprint JSON data, meaning we just have to special case this. I imagine there are more custom fields like this as well unfortunately :/

Reply:
i don't think i'm grokking this properly, we have special case logic if the schema type is `array` but the field is not `sprint`?

Reply:
Kind of. We need to iterate over any field with type `array` and run the sub-transformers on each item if we're passed an iterable. `sprint` custom fields are funny in that Jira errors if we pass them a list, which is expected by every other `array` type I've tested, so by excluding it from the iteration logic here, it falls through to the custom type handler on L56.

Reply:
As an alternative to make this a little less messy, I could move this logic in to an array type transformer that calls into the valid sub-type transformer 🤔 It just becomes a recursive loop if we ever receive a schema with type: `array`, items: `array`.

Reply:
ohhh we're ignoring the sprint field here. got it